conans/model/scm.py

```diff
@@ -46,10 +63,22 @@ def recipe_revision(self):
 
     def __repr__(self):
         d = {"url": self.url, "revision": self.revision, "username": self.username,
-             "password": self.password, "type": self.type, "verify_ssl": self.verify_ssl,
+             "password": self.password, "type": self.type,
              "subfolder": self.subfolder, "submodule": self.submodule}
+        if self.shallow != self.SHALLOW_DEFAULT:
+            d.update({"shallow": self.shallow})
+        if self.verify_ssl != self.VERIFY_SSL_DEFAULT:
+            d.update({"verify_ssl": self.verify_ssl})
         d = {k: v for k, v in d.items() if v is not None}
-        return json.dumps(d, sort_keys=True)
+
+        def _kv_to_string(key, value):
+            if isinstance(value, bool):
+                return '"{}": {}'.format(key, value)
+            else:
+                value_str = str(value).replace('"', r'\"')
+                return '"{}": "{}"'.format(key, value_str)
```

Comment:
`NoneType` and `int` will be quoted too?

Reply:
* Entries with `None` values are removed from the dictionary.
 * `int` will be quoted too, but IMO using `int` for `revision` (the only field that accepts it) should be discouraged too (it makes sense only for SVN)

Reply:
Then I think we should just transform to String anything it comes (even a function) and simplify the checks by checking only the Boolean fields (the SSL and the shallow).

Reply:
The checking code seems correct to me but not short nor simple, so, maybe it is too much.

Reply:
I don't think it will simplify a lot the code... it would be more or like this:

```diff
diff --git a/conans/model/scm.py b/conans/model/scm.py
index 172cf8a3a..c62fcc101 100644
--- a/conans/model/scm.py
+++ b/conans/model/scm.py
@@ -16,17 +16,17 @@ def get_scm_data(conanfile):
         return None
 
 
-def _get_dict_value(data, key, expected_type, default=None, disallowed_type=None):
+def _get_dict_value(data, key, expected_type=None, default=None):
     if key in data:
         r = data.get(key)
         if r is None:  # None is always a valid value
             return r
-        if not isinstance(r, expected_type) or (disallowed_type and isinstance(r, disallowed_type)):
+        if expected_type and not isinstance(r, expected_type):
             type_str = "' or '".join([it.__name__ for it in expected_type]) \
                 if isinstance(expected_type, tuple) else expected_type.__name__
             raise ConanException("SCM value for '{}' must be of type '{}'"
                                  " (found '{}')".format(key, type_str, type(r).__name__))
-        return r
+        return str(r) if expected_type is not bool else r
     return default
 
 
@@ -36,15 +36,14 @@ class SCMData(object):
 
     def __init__(self, conanfile):
         data = getattr(conanfile, "scm")
-        self.type = _get_dict_value(data, "type", string_types)
-        self.url = _get_dict_value(data, "url", string_types)
-        self.revision = _get_dict_value(data, "revision", string_types + (int, ),
-                                        disallowed_type=bool)  # bool is subclass of integer
+        self.type = _get_dict_value(data, "type")
+        self.url = _get_dict_value(data, "url")
+        self.revision = _get_dict_value(data, "revision")
         self.verify_ssl = _get_dict_value(data, "verify_ssl", bool, SCMData.VERIFY_SSL_DEFAULT)
-        self.username = _get_dict_value(data, "username", string_types)
-        self.password = _get_dict_value(data, "password", string_types)
-        self.subfolder = _get_dict_value(data, "subfolder", string_types)
-        self.submodule = _get_dict_value(data, "submodule", string_types)
+        self.username = _get_dict_value(data, "username")
+        self.password = _get_dict_value(data, "password")
+        self.subfolder = _get_dict_value(data, "subfolder")
+        self.submodule = _get_dict_value(data, "submodule")
         self.shallow = _get_dict_value(data, "shallow", bool, SCMData.SHALLOW_DEFAULT)
 
     @property
```

The drawback is that it will be converting to string (without errors) things that previously could be any type. **Anyway, the only legit issue I can think of** is the integer for an SVN revision, but I think it could be very marginal.

Shall I push these changes then? wdyt @memsharded?